src/main/src/main/java/org/geoserver/security/decorators/SecuredObjects.java

```diff
@@ -45,9 +46,13 @@ public static Object secure(Object object, WrapperPolicy policy) {
         if (object == null)
             return null;
 
+        // In case the object to be secured is wrapped in a ModificationProxy, get the proxied Object class
+        // if the object is not wrapped, ModificationProxy.unwrap() will just return the object reference.
+        final Object unwrappedObject = ModificationProxy.unwrap(object);
+
         // if we already know what can handle the wrapping, just do it, don't
         // scan the extension points once more
-        Class<?> clazz = object.getClass();
+        Class<?> clazz = unwrappedObject.getClass();
```

Comment:
Seems fine but can you elaborate on this change?

Reply:
Sure thing.

In debugging the issue, I noticed that the code here was handling dynamic Proxy object instances (`java.lang.reflect.Proxy.$Proxy101` for example) in some cases, and normal Catalog objects in other cases (`org.geoserver.catalog.impl.DataStoreInfoImpl` for example), where the Proxy objects were ModificationProxy objects wrapping _normal_ Catalog objects. While the Proxy objects will still pass the `canSecure()` method as long as they wrap an object instance of a supported type, it seemed possible for the Factory cache to grow larger than necessary over time, if the Proxy instances changed. I felt it was safer to use the unwrapped class type as the key for this cache, to avoid the possibility of multiple Proxy instances wrapping the same underlying type being added to the cache.

I'n debugging through a running instance, I didn't observe multiple Proxy instances wrapping the same type come through this code, however. So I don't know how possible or impossible it actually is over time. But it seems safer to have the actual Catalog class types in the cache.